node/pkg/driver/node.go

```diff
@@ -613,9 +619,75 @@ func (d *NodeService) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 }
 
 func (d *NodeService) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	goid_info.SetAdditionalIDInfo(req.VolumeId)
+	volumeId := req.VolumeId
+	goid_info.SetAdditionalIDInfo(volumeId)
 	defer goid_info.DeleteAdditionalIDInfo()
-	return nil, status.Error(codes.Unimplemented, "NodeGetVolumeStats is not implemented yet")
+	volumePath := req.VolumePath
+	volumePathWithHostPrefix := d.NodeUtils.GetPodPath(volumePath)
+
+	err := d.nodeGetVolumeStatsRequestValidation(volumeId, volumePath)
+	if err != nil {
+		return nil, err
+	}
+
+	isPathExists := d.NodeUtils.IsPathExists(volumePathWithHostPrefix)
+	if !isPathExists {
+		return nil, status.Errorf(codes.NotFound, "volume path %q does not exist", volumePath)
+	}
+
+	volumeStats, err := d.getVolumeStats(volumePathWithHostPrefix, volumeId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to get statistics for volume path %q: %s", volumePath, err)
+	}
+	return &csi.NodeGetVolumeStatsResponse{
+		Usage: []*csi.VolumeUsage{
+			{
+				Unit:      csi.VolumeUsage_BYTES,
+				Available: volumeStats.AvailableBytes,
+				Total:     volumeStats.TotalBytes,
+				Used:      volumeStats.UsedBytes,
+			},
+			{
+				Unit:      csi.VolumeUsage_INODES,
+				Available: volumeStats.AvailableInodes,
+				Total:     volumeStats.TotalInodes,
+				Used:      volumeStats.UsedInodes,
+			},
+		},
+	}, nil
+}
+
+func (d *NodeService) nodeGetVolumeStatsRequestValidation(volumeId string, volumePath string) error {
+	if volumeId == "" {
+		return status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+	}
+	if volumePath == "" {
+		return status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume Path must be provided")
+	}
+
+	return nil
+}
+
+func (d *NodeService) getVolumeStats(path string, volumeId string) (VolumeStatistics, error) {
+	isBlock, err := d.NodeUtils.IsBlock(path)
+	if err != nil {
+		return VolumeStatistics{}, status.Errorf(codes.Internal, "failed to determine if %q is block device: %s", path, err)
+	}
+
+	if isBlock {
+		mpathDevice, err := d.OsDeviceConnectivityHelper.GetMpathDevice(volumeId)
+		if err != nil {
+			switch err.(type) {
+			case *device_connectivity.MultipathDeviceNotFoundForVolumeError:
+				return VolumeStatistics{}, status.Errorf(codes.NotFound, "Multipath Device of volume id %q does not exist", volumeId)
```

Comment:
wouldn't kubelet still get an internal error [here](https://github.com/IBM/ibm-block-csi-driver/blob/00ee18991eb5c624f26c6c54f8b694fc98d88be5/node/pkg/driver/node.go#L640)?

Reply:
yes you right, I will move this function to the `NodeGetVolumeStats` function

Reply:
another option is to have `getVolumeStats` return only grpc errors, so that `NodeGetVolumeStats` can simply return the error

Reply:
yes yes I thought about it later and wanted to do like you suggested

Reply:
done